### PR TITLE
webgpu: optimize matmul with small output size

### PR DESCRIPTION
**Pull request by xhcao**
This patch changes to read data from global memory to registers,
then store registers' content to shared memory, so it is
instruction-Level parallelism to avoid to need to two buffers
and reduce by half the number of block threads.

To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6501)

### CODE REVIEW

**Comment by xhcao** [comment]
I had run the bodypix model with 0.25 | ResNet50 | tensor | webgpu configuration on Tiger-Lake machine, and got the performance data as below,
<html>
<body>
<b style="font-weight:normal;" id="docs-internal-guid-3673657b-7fff-3836-174f-b5453b5f43f8"><div dir="ltr" style="margin-left:0pt;" align="left">

M, K, N | Without the patch | With the patch
-- | -- | --
2, 2048, 512 | 0.315 | 0.211
2, 2048, 48 | 0.256 | 0.158
2, 2048, 34 | 0.258 | 0.156
2, 2048, 32 | 0.258 | 0.147
2, 2048, 17 | 0.258 | 0.146
2, 2048, 24 | 0.257 | 0.145
2, 1024, 512 | 0.187 | 0.117
3, 1024, 256 | 0.142 | 0.095
5, 512, 128 | 0.081 | 0.048
3, 512, 256 | 0.076 | 0.048
5, 256, 512 | 0.054 | 0.034
9, 256, 64 | 0.040 | 0.028
5, 256, 128 | 0.044 | 0.027
5, 128, 512 | 0.034 | 0.025
3, 128, 512 | 0.028 | 0.023
9, 64, 256 | 0.017 | 0.015
5, 64, 256 | 0.020 | 0.015
9, 64, 64 | 0.018 | 0.014

</div></b>
</body>
</html>

**Comment by xhcao** [comment]
@qjia7 please take a look, thank you.

**Comment by qjia7** [comment]
In your change, `makeMatMulSmallOutputSizeSource` is very similar with [`makeMatMulPackedSource`](https://github.com/tensorflow/tfjs/blob/master/tfjs-backend-webgpu/src/matmul_packed_webgpu.ts#L46) with workGroupSize =  [16, 8, 1] and elementsPerThread = [1, 1, 1] except missing the step that read data from global memory to registers. In this case, can you directly use `makeMatMulPackedSource` and config the right workGroupSize and work per thread size based on input shapes?

**Comment by xhcao** [comment]
Hi, @qjia7 , The `MatMulSmallOutputSizeProgram` previously used double shared memory buffers to ensure ALU arithmetic operations and global memory load simultaneously in the same barrier domain, but a half threads execute arithmetic operations and the other half threads load global memory in one work group, actually, the commands are instruction-Level parallelism, so we could execute arithmetic operations and global memory load in the same threads, this patch removes the redundant threads.
In order to balance the shared memory and registers resources, the patch uses the registers instead of a shared memory buffer, because it is very quick to copy data from shared memory to registers, the patch keeps the advantage of double buffers here, and the code is greatly simplified.
For small size matmul, the mumber of work groups is also small, so we explicitly use registers here. If we introduce it to `MatMulPackedProgram` and the mumber of work groups is also large, we may get worse performance for register overflow issue.
I had collected data the bodypix model with 0.25 | ResNet50 | tensor | webgpu configuration on Tiger-Lake machine again, and shown it as below, the last column is  workGroupSize = [16, 8, 1] and elementsPerThread = [1, 1, 1].

<html>
<body>
<b style="font-weight:normal;" id="docs-internal-guid-962382e9-7fff-5c66-d9d7-049ea3460245"><div dir="ltr" style="margin-left:0pt;" align="left">

M, K, N | MatMulSmallOutputSizeProgram without the CL | MatMulSmallOutputSizeProgram with the CL | Configured MatMulPackedProgram
-- | -- | -- | --
2, 2048, 512 | 0.315 | 0.212 | 0.280
2, 2048, 48 | 0.256 | 0.150 | 0.230
2, 2048, 34 | 0.278 | 0.152 | 0.211
2, 2048, 32 | 0.257 | 0.141 | 0.231
2, 2048, 17 | 0.273 | 0.156 | 0.195
2, 2048, 24 | 0.257 | 0.140 | 0.208
2, 1024, 512 | 0.162 | 0.115 | 0.138
3, 1024, 256 | 0.145 | 0.088 | 0.130
5, 512, 128 | 0.082 | 0.046 | 0.072
3, 512, 256 | 0.076 | 0.047 | 0.067
5, 256, 512 | 0.057 | 0.035 | 0.045
9, 256, 64 | 0.040 | 0.028 | 0.037
5, 256, 128 | 0.043 | 0.028 | 0.038
5, 128, 512 | 0.035 | 0.024 | 0.030
3, 128, 512 | 0.028 | 0.022 | 0.044
9, 64, 256 | 0.017 | 0.019 | 0.016
5, 64, 256 | 0.020 | 0.015 | 0.016
9, 64, 64 | 0.018 | 0.014 | 0.016

</div><br /></b>
</body>
</html>

**Comment by qjia7** [comment]
Good analysis, Xinghua. But please also check `Configured MatMulPackedVec4Program`. I believe `Configured MatMulPackedVec4Program` will be better than `Configured MatMulPackedProgram`. I want to know the necessary to keep the `MatMulSmallOutputSizeProgram with the CL`.

**Comment by xhcao** [comment]
Hi, I had configured `MatMulPackedVec4Program` and not got better performance. `MatMulPackedVec4Program` is nearly similar as `MatMulPackedProgram`, which is worse than `MatMulSmallOutputSizeProgram `.

[1] -- MatMulSmallOutputSizeProgram without the patch
[2] -- MatMulSmallOutputSizeProgram with the patch
[3] -- Configured MatMulPackedProgram
[4] -- Configured MatMulPackedVec4Program
<html>
<body>
<b style="font-weight:normal;" id="docs-internal-guid-5ae5d7d3-7fff-4b33-12ed-1d218d10027c"><div dir="ltr" style="margin-left:0pt;" align="left">

M, K, N | [1] | [2] | [3] | [4]
-- | -- | -- | -- | --
2, 2048, 512 | 0.315 | 0.212 | 0.280 | 0.296
2, 2048, 48 | 0.256 | 0.150 | 0.230 | 0.277
2, 2048, 34 | 0.278 | 0.152 | 0.211 | --
2, 2048, 32 | 0.257 | 0.141 | 0.231 | 0.242
2, 2048, 17 | 0.273 | 0.156 | 0.195 | --
2, 2048, 24 | 0.257 | 0.140 | 0.208 | 0.270
2, 1024, 512 | 0.162 | 0.115 | 0.138 | 0.146
3, 1024, 256 | 0.145 | 0.088 | 0.130 | 0.133
5, 512, 128 | 0.082 | 0.046 | 0.072 | 0.078
3, 512, 256 | 0.076 | 0.047 | 0.067 | 0.073
5, 256, 512 | 0.057 | 0.035 | 0.045 | 0.044
9, 256, 64 | 0.040 | 0.028 | 0.037 | 0.041
5, 256, 128 | 0.043 | 0.028 | 0.038 | 0.043
5, 128, 512 | 0.035 | 0.024 | 0.030 | 0.029
3, 128, 512 | 0.028 | 0.022 | 0.044 | 0.028
9, 64, 256 | 0.017 | 0.019 | 0.016 | 0.017
5, 64, 256 | 0.020 | 0.015 | 0.016 | 0.018
9, 64, 64 | 0.018 | 0.014 | 0.016 | 0.019

</div></b>
</body>
</html>

File: tfjs-backend-webgpu/src/matmul_small_output_size_webgpu.ts [hunk 1]
```diff
@@ -143,7 +105,7 @@ export class MatMulSmallOutputSizeProgram implements WebGPUProgram {
     this.dispatchLayout = {x: [2], y: [1], z: [0]};
     this.dispatch = [
       Math.ceil(outputShape[2] / this.workGroupSize[0]),
-      Math.ceil(outputShape[1] * 2 / this.workGroupSize[1]), outputShape[0]
+      Math.ceil(outputShape[1] / this.workGroupSize[1]), outputShape[0]
     ];
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/matmul_small_output_size_webgpu.ts:109]: Nit: Please directly use
 `this.dispatch = computeDispatch( this.dispatchLayout, this.outputShape, this.workGroupSize);`

**Review by qjia7 — APPROVED** [review]
Thanks Xinghua for the detailed testing!

**Review by gyagp — APPROVED** [review]
LGTM. Thanks for the optimization!

**Pull request merged**